### PR TITLE
feat: add endpoint for override plans

### PR DESCRIPTION
**Pull request by lovrocolic**
## Context

This PR add all necessary logic to perform plan override through the API

### CODE REVIEW

File: app/controllers/api/v1/subscriptions_controller.rb [hunk 1]
```diff
@@ -22,6 +22,31 @@ def create
...
+        plan_params = Plans::PrepareForOverrideService
+          .new(current_organization, overridden_plan_code_params[:overridden_plan_code])
+          .call(plan_params: create_with_override_plan_params)
+
+        override_service = Subscriptions::OverrideService.new
+
+        result = override_service.call_from_api(
+          organization: current_organization,
+          plan_args: plan_params,
+          subscription_args: create_with_override_subscription_params
```
Comment (jdenquin) [inline app/controllers/api/v1/subscriptions_controller.rb:35]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: app/services/plans/prepare_for_override_service.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+      plan
+    end
+
+    private
+
+    attr_reader :organization, :plan_code
+
+    def overridden_plan
+      @overridden_plan ||= begin
```
Comment (jdenquin) [inline app/services/plans/prepare_for_override_service.rb:38]: Style/RedundantBegin: Redundant `begin` block detected. (https://rubystyle.guide#begin-implicit)

File: spec/requests/api/v1/subscriptions_spec.rb [hunk 3]
```diff
@@ -51,6 +51,105 @@
...
+                {
+                  to_value: nil,
+                  from_value: 2,
+                  flat_amount: '0',
+                  per_unit_amount: '3200',
+                },
+              ],
+            },
+          ],
+        }
```
Comment (jdenquin) [inline spec/requests/api/v1/subscriptions_spec.rb:101]: Style/TrailingCommaInHashLiteral: Put a comma after the last item of a multiline hash.

File: spec/requests/api/v1/subscriptions_spec.rb [hunk 4]
```diff
@@ -51,6 +51,105 @@
...
+      }
+    end
+
+    before { plan }
+
+    it 'returns a success' do
+      post_with_token(organization, '/api/v1/subscriptions/override', { subscription: params })
+
+      expect(response).to have_http_status(200)
```
Comment (jdenquin) [inline spec/requests/api/v1/subscriptions_spec.rb:110]: RSpec/Rails/HttpStatus: Prefer `:ok` over `200` to describe HTTP status code. (https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/Rails/HttpStatus)

File: spec/requests/api/v1/subscriptions_spec.rb [hunk 5]
```diff
@@ -51,6 +51,105 @@
...
+      let(:params) do
+        {
+          external_customer_id: customer.external_id,
+          name: 'subscription name',
+          overridden_plan_code: plan.code,
+          external_id: SecureRandom.uuid,
+          billing_time: 'anniversary',
+          plan: {
+            amount_currency: 'EUR',
+          }
```
Comment (jdenquin) [inline spec/requests/api/v1/subscriptions_spec.rb:141]: Style/TrailingCommaInHashLiteral: Put a comma after the last item of a multiline hash.

File: spec/requests/api/v1/subscriptions_spec.rb [hunk 6]
```diff
@@ -51,6 +51,105 @@
...
+        }
+      end
+
+      it 'returns an unprocessable_entity error' do
+        post_with_token(organization, '/api/v1/subscriptions/override', { subscription: params })
+
+        expect(response).to have_http_status(422)
```
Comment (jdenquin) [inline spec/requests/api/v1/subscriptions_spec.rb:148]: RSpec/Rails/HttpStatus: Prefer `:unprocessable_entity` over `422` to describe HTTP status code. (https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/Rails/HttpStatus)

File: spec/services/plans/prepare_for_override_service_spec.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+      amount_cents: 200,
+      amount_currency: 'EUR',
+      charges: [
+        {
+          id: standard_charge.id,
+          charge_model: 'standard',
+          properties: {
+            amount: '0.22',
+          },
+        }
```
Comment (jdenquin) [inline spec/services/plans/prepare_for_override_service_spec.rb:25]: Style/TrailingCommaInArrayLiteral: Put a comma after the last item of a multiline array. (https://rubystyle.guide#no-trailing-array-commas)

File: spec/services/plans/prepare_for_override_service_spec.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+      ]
```
Comment (jdenquin) [inline spec/services/plans/prepare_for_override_service_spec.rb:26]: Style/TrailingCommaInHashLiteral: Put a comma after the last item of a multiline hash.

File: spec/services/plans/prepare_for_override_service_spec.rb [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+      amount_cents: 200,
+      amount_currency: 'EUR',
+      charges: [
+        {
+          billable_metric_id: standard_charge.billable_metric_id,
+          charge_model: 'standard',
+          properties: {
+            amount: '0.22',
+          },
+        }
```
Comment (jdenquin) [inline spec/services/plans/prepare_for_override_service_spec.rb:49]: Style/TrailingCommaInArrayLiteral: Put a comma after the last item of a multiline array. (https://rubystyle.guide#no-trailing-array-commas)

File: spec/services/subscriptions/override_service_spec.rb [hunk 10]
```diff
@@ -120,4 +121,119 @@
...
+      }
+    end
+
+    before { plan }
+
+    it 'creates a subscription' do
+      result = override_service.call_from_api(
+        organization: organization,
+        plan_args: plan_params,
+        subscription_args: subscription_params
```
Comment (jdenquin) [inline spec/services/subscriptions/override_service_spec.rb:165]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: spec/services/subscriptions/override_service_spec.rb [hunk 11]
```diff
@@ -120,4 +121,119 @@
...
+        expect(subscription).to be_anniversary
+      end
+    end
+
+    it 'creates a new plan' do
+      expect do
+        override_service.call_from_api(
+          organization: organization,
+          plan_args: plan_params,
+          subscription_args: subscription_params
```
Comment (jdenquin) [inline spec/services/subscriptions/override_service_spec.rb:187]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: spec/services/subscriptions/override_service_spec.rb [hunk 12]
```diff
@@ -120,4 +121,119 @@
...
+        )
+      end.to change(Plan, :count).by(1)
+    end
+
+    it 'creates a new plan that is correctly linked to overridden_plan' do
+      result = override_service.call_from_api(
+        organization: organization,
+        plan_args: plan_params,
+        subscription_args: subscription_params
```
Comment (jdenquin) [inline spec/services/subscriptions/override_service_spec.rb:196]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: spec/services/subscriptions/override_service_spec.rb [hunk 13]
```diff
@@ -120,4 +121,119 @@
...
+    end
+
+    context 'when plan is invalid' do
+      let(:plan_name) { nil }
+
+      it 'fails' do
+        result = override_service.call_from_api(
+          organization: organization,
+          plan_args: plan_params,
+          subscription_args: subscription_params
```
Comment (jdenquin) [inline spec/services/subscriptions/override_service_spec.rb:209]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: spec/services/subscriptions/override_service_spec.rb [hunk 14]
```diff
@@ -120,4 +121,119 @@
...
+    end
+
+    context 'when billing_time is invalid' do
+      let(:billing_time) { :foo }
+
+      it 'fails' do
+        result = override_service.call_from_api(
+          organization: organization,
+          plan_args: plan_params,
+          subscription_args: subscription_params
```
Comment (jdenquin) [inline spec/services/subscriptions/override_service_spec.rb:227]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

**Comment by lovrocolic** [comment]
This feature will not be used for now

**Pull request closed without merging**